bindings/node/test/autoEncrypter.test.js

```diff
@@ -2,10 +2,10 @@
 
 const fs = require('fs');
 const path = require('path');
-const BSON = require('bson');
-const EJSON = require('bson').EJSON;
 const sinon = require('sinon');
 const mongodb = require('mongodb');
+const BSON = mongodb.BSON;
```

Comment:
Did you run the tests locally with `SKIP_LIVE_TESTS` set to false?  My hunch is that tests will fail.  

The mongodb dev dependency is 4.8.1, which doesn't export bson.  Some of the tests (which I believe are skipped when there is no server running, like in CI) create a mock client using the `BSON` variable and store it on the topology.

For the tests, `BSON` is undefined (because we're running with 4.8.1) so my guess is that those tests will fail and CI is currently passing because the tests are skipped.

Reply:
The tests failing is expected and is my last comment in the description of the PR:

_The patch build shows this works (since the tests pass and bson is no longer a dev dependency) but the dev dependency on mongodb will need to bump to the next release once https://github.com/mongodb/node-mongodb-native/pull/3367 is merged._

Reply:
As long as the intent is to always run latest, I suppose this is fine. But in that case I'm a little concerned that we don't have any new tests explicitly testing the fallback order. I'm also wondering why we *don't* want to check for the passed in bson options first and only fall back to the exported definition last, since if it's defined, it should be fine, right? Or is this just to indicate that the standard moving forward will be the driver exported library and that we'll be getting rid of the ability to pass it in via options or reading it from the topology altogether?

I just wish we had a more clear documentation (via test) about the 3 different fallback cases.

Reply:
P.S. should we bump the dependency in this PR after we release 4.9.0?

Reply:
@dariakp Yes we should update to 4.9.0 but it won't make much difference as the live tests aren't running in this repo as @baileympearson pointed out. I updated the pinned commit in the Node driver to point at the latest commit in this PR, so when this merges the Node driver tests will cover the changes. I updated the description of this PR and here's again a link to that run passing:

https://spruce.mongodb.com/task/mongo_node_driver_next_ubuntu1804_custom_dependency_tests_run_custom_csfle_tests_latest_pinned_commit_40d485c292a14f8251a4e3724d1bb5f8cf11577c_22_08_17_20_53_00/tests?execution=0&sortBy=STATUS&sortDir=ASC

Reply:
Ok, but what about my other question above? (Regarding actually testing the three different fallback cases)

> As long as the intent is to always run latest, I suppose this is fine. But in that case I'm a little concerned that we don't have any new tests explicitly testing the fallback order. I'm also wondering why we don't want to check for the passed in bson options first and only fall back to the exported definition last, since if it's defined, it should be fine, right? Or is this just to indicate that the standard moving forward will be the driver exported library and that we'll be getting rid of the ability to pass it in via options or reading it from the topology altogether?
>
> I just wish we had a more clear documentation (via test) about the 3 different fallback cases.

Reply:
Yeah, that's a good point.  Users who do upgrade to version 4.9.0 of the driver and mongodb-client-encryption@2.2.1 _and_ pass in bson as an option would find that suddenly mongodb-client-encryption doesn't use the bson library they provided anymore.  Technically that's a breaking change? 

As far as the ordering goes - it needs to check the topology last.  Otherwise, mongodb-client-encryption will always check the topology, even if the client isn't connected.  But I do believe the intent was that eventually we'd stop supporting the BSON option too and enforce that mongodb-client-encryption and the driver always have the same bson version (@nbbeeken @durran right?).  So I think the order should be options, BSON import, topology.

Reply:
I've updated to use options first, then BSON, then topology.

Reply:
@baileympearson @dariakp I've updated the unit tests to test all the bson permutations.